File: leetcode/algorithms/heap/kth-largest-element-in-a-stream.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Kth Largest Element in a Stream
*/

class MinHeapIntegerComparator implements Comparator<Integer> {
    public int compare(Integer i1, Integer i2){
        return i1 - i2;
    }
}

class KthLargest {
    
    PriorityQueue<Integer> minHeap = null;
    Integer cap = null;
    public KthLargest(int k, int[] nums) {
        cap = k;
        minHeap = new PriorityQueue<Integer>(new MinHeapIntegerComparator());
        for(int num : nums){
            minHeap.add(num);
            if(minHeap.size()>cap){
                minHeap.poll();
            }
        }
    }
    
    public int add(int val) {
        minHeap.add(val);
        if(minHeap.size()>cap){
            minHeap.poll();
        }
        return minHeap.peek();
    }
}

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest obj = new KthLargest(k, nums);
 * int param_1 = obj.add(val);
 */

File: leetcode/algorithms/heap/kth-largest-element-in-a-stream.py
class KthLargest(object):

    def __init__(self, k, nums):
        """
        :type k: int
        :type nums: List[int]
        """
        self.k=k
        self.kheap = []
        heapq.heapify(self.kheap)
        
        for i in nums:
            self.add(i)
        

    def add(self, val):
        """
        :type val: int
        :rtype: int
        """
        heapq.heappush(self.kheap,val)
        if len(self.kheap)>self.k:
                heapq.heappop(self.kheap)
        return self.kheap[0]
        
        


# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)